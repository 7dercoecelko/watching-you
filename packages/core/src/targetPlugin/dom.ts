import { Target } from '../type/Target';
import { round } from '../util/math';

class DomTarget implements Target {
  name = 'dom';
  target: HTMLElement | null = null;

  constructor(target: HTMLElement | string) {
    try {
      if (typeof target === 'string') {
        this.target = document.querySelector(target);
        return;
      }
      this.target = target;
    } catch (e) {
      this.target = null;
    }
  }
  setup() {}
  cleanup() {}
  update() {
    if (!this.target) return { x: 0, y: 0 };

    const rect = this.target.getBoundingClientRect();

    // XXX: Only return the center position of the dom now
    const x = round(rect.left + rect.width / 2);
    const y = round(rect.top + rect.height / 2);
    return { x, y };
  }
}

export default DomTarget;
