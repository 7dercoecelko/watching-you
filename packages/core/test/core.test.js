/**
 * @vitest-environment happy-dom
 */
import { beforeEach, describe, expect, it } from 'vitest';
import { fireEvent } from '@testing-library/dom';
import WatchingYou from '../src/index';

const wait = async (ms) => {
  await new Promise((r) => {
    setTimeout(r,ms);
  });
};

const mousemove = async (x, y) => {
  fireEvent(
    window,
    new MouseEvent('mousemove', {
      clientX: x,
      clientY: y,
    }),
  );
  await window.happyDOM.whenAsyncComplete();
  await wait(100);
};

describe('WatchingYou core test', () => {
  beforeEach(() => {
    document.body.innerHTML = `
      <div class="eyes">
        <div>O</div>
      </div>
    `;
  });

  it('It should be able to be initialized', () => {
    const watchingYou = new WatchingYou();
    expect(watchingYou instanceof WatchingYou).toBe(true);
  });

  it('It should be watching mouse', async () => {
    const eyesDom = document.querySelector('.eyes');
    const watchingYou = new WatchingYou(eyesDom);
    watchingYou.start();
    await mousemove(100, 200);
    expect(eyesDom.style.transform).not.toBe('none');
    expect(eyesDom.style.transform).not.toBe(
      'translate(0px,0px) rotate(0deg)',
    );
    expect(eyesDom.style.transform).toMatchSnapshot();
    await mousemove(300, 300);
    expect(eyesDom.style.transform).toMatchSnapshot();
  });
});
