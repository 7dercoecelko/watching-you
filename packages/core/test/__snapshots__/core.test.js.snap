// Vitest Snapshot v1

exports[`WatchingYou core test > It should be watching mouse 1`] = `"translate3d(29.3px,40.5px,0px) rotate(144.1deg)"`;

exports[`WatchingYou core test > It should be watching mouse 2`] = `"translate3d(37.2px,33.5px,0px) rotate(132deg)"`;
